rfc/20250730-module-misc-settings.md

```diff
@@ -0,0 +1,429 @@
+    for them for now, until we have a better idea of whether and how we might
+    use them in OpenTofu.
+
+    This proposal includes them primarily because thematically they seem to
+    belong to the same category of settings as the OpenTofu runtime version
+    constraints and so proposing a new location for all of them together
+    seemed wise. However, we could choose to introduce the new `language`
+    block type with only the `compatible_with` block type to start and
+    then add other arguments later once we know what problems we're trying
+    to solve, which would give us more freedom to choose to do something
+    significantly different than what's stubbed today.
+    
+    The only slight advantage of doing something for these immediately is that
+    -- as with the existing language features for these concepts -- we can
+    introduce real uses of them later knowing that at least some older versions
+    of OpenTofu recognize them enough to return a specialized error message
+    about them. However, we could achieve that a different way by ensuring
+    that the version constraints in `compatible_with` are always checked
+    before returning any other errors and then expect that module authors
+    using whatever hypothetical language-edition-like or experiment-like
+    features we add will _also_ use `compatible_with` to exclude OpenTofu
+    versions that do not support the new arguments.
+
+    There is also a potential compromise in supporting the `experiments`
+    argument as an alias for the existing argument of the same name but not
+    including `edition` at all. There is already existing logic in OpenTofu
+    to handle `experiments`, but our only handling of language editions today
+    is to return an error if the argument is set to anything other than a
+    placeholder token representing an older version of the Terraform language.
+
+- **Should we allow modules to specify that they aren't compatible with OpenTofu _at all_?**
+
+    The current proposal focuses on the situation where a module is written
+    to support OpenTofu but wants to declare that it's only intended to work
+    with a certain subset of OpenTofu versions.
+
+    It does not include any way to assert that the module is not intended to
+    be used with _any_ version of OpenTofu, i.e. that it is intended for use
+    only with Terraform or with some other hypothetical future tool that
+    replaces OpenTofu while still supporting its module format.
+
+    We could potentially support a special extra value assigned to `opentofu`
+    in a `compatible_with` block which represents an empty set of compatible
+    versions, whereas the default when unspecified is the maximum set containing
+    _all_ possible versions. Is this useful enough to be worth the additional
+    complexity that implies?
+
+    (Technically we could add such a thing later as long as earlier versions
```

Comment:
It's not as " builtin", but a user could set an invalid opentofu requirement.

This can be accomplished today with a main.tofu file with `required_version = "unsupported"`, which results in
```
╷
│ Error: Invalid version constraint
│ 
│   on main.tofu line 2, in terraform:
│    2:         required_version = "unsupported"
│ 
│ This string does not use correct version constraint syntax.
```

Reply:
Indeed, that's basically where this meandering parenthetical landed:

> If that worse error message were acceptable then the author could just set the version constraint to any invalid version constraint syntax to get the same effect.

...so the essence of this open question is: should we just consider this a rare enough need that it's okay to rely on just making the configuration invalid from OpenTofu's perspective in this way, at the expense of a worse error message?

I found myself leaning towards accepting this compromise as I was writing it out, since I don't expect this would be a very common thing to need to express.

Reply:
I think that sounds reasonable.  I can leave this conversation unresolved to see if there's any other thoughts on it.